featurebyte/feast/infra/offline_stores/spark_thrift.py

```diff
@@ -0,0 +1,353 @@
+import pandas as pd
+import pyarrow
+from feast import FeatureView, OnDemandFeatureView
+from feast.data_source import DataSource
+from feast.feature_view import DUMMY_ENTITY_ID, DUMMY_ENTITY_VAL
+from feast.infra.offline_stores.offline_store import OfflineStore, RetrievalJob, RetrievalMetadata
+from feast.infra.registry.base_registry import BaseRegistry
+from feast.repo_config import FeastConfigBaseModel, RepoConfig
+from feast.saved_dataset import SavedDatasetStorage
+from pytz import utc
+from sqlglot import expressions
+
+from featurebyte.enum import SourceType, StorageType
+from featurebyte.feast.infra.offline_stores.spark_thrift_source import SparkThriftSource
+from featurebyte.logging import get_logger
+from featurebyte.models.credential import StorageCredential
+from featurebyte.query_graph.sql.common import quoted_identifier, sql_to_string
+from featurebyte.session.base_spark import BaseSparkSession
+from featurebyte.session.spark import SparkDatabaseCredential, SparkSession
+
+logger = get_logger(__name__)
+
+
+class BaseSparkThriftOfflineStoreConfig(FeastConfigBaseModel):
+    """Base Offline store config for Spark Thrift Server"""
+
+    host: str
+    http_path: str
+    storage_path: str
+    catalog_name: str
+    schema_name: str
+
+    @abstractmethod
+    def get_db_session(self) -> BaseSparkSession:
+        """
+        Get a Spark Thrift Session
+
+        Returns
+        -------
+        BaseSparkSession
+            Spark Thrift Session
+
+        Raises
+        ------
+        ValueError
+            If the store_config is not a valid Spark Thrift Offline Store Config
+        """
+
+    def get_table_column_names_and_types(
+        self,
+        table_name: str,
+    ) -> Iterable[Tuple[str, str]]:
+        """
+        Get the column names and types of a table
+
+        Parameters
+        ----------
+        table_name: str
+            Table name
+
+        Returns
+        -------
+        Iterable[Tuple[str, str]]
+        """
+        db_session = self.get_db_session()
+        result = db_session.execute_query_blocking(f"DESCRIBE {table_name}")
+        assert result is not None
+        return tuple((row.col_name, row.data_type) for _, row in result.iterrows())
+
+
+class SparkThriftOfflineStoreConfig(BaseSparkThriftOfflineStoreConfig):
+    """Offline store config for Spark Thrift Server"""
+
+    type: Literal["spark_thrift.offline"] = "spark_thrift.offline"
+    """ Offline store type selector"""
+
+    port: int
+    use_http_transport: bool
+    use_ssl: bool
+    storage_type: StorageType
+    storage_url: str
+    database_credential: Optional[SparkDatabaseCredential]
+    storage_credential: Optional[StorageCredential]
+
+    def get_db_session(self) -> BaseSparkSession:
+        return SparkSession(**self.dict())
+
+
+class SparkThriftOfflineStore(OfflineStore):
+    """
+    Offline store for Spark Thrift Server
+    """
+
+    @staticmethod
+    def pull_latest_from_table_or_query(
```

Comment:
Curious what are the differences between the query generated by this vs the spark offline store? https://github.com/feast-dev/feast/blob/9a3590ea771ca3c3224f5e1a833453144e54284e/sdk/python/feast/infra/offline_stores/contrib/spark_offline_store/spark.py#L59-L62

Reply:
Oh it's the dependency on pyspark. The query probably is quite similar.

Reply:
Yes the query is derived from there but implemented using sqlglot for better robustness and to ensure proper quoting